src/axom/inlet/Inlet.hpp

```diff
@@ -543,6 +578,22 @@ class Inlet
     return m_globalTable.addGenericDictionary(name, description);
   }
 
+  /*!
+   *****************************************************************************
+   * \brief Add an dictionary of user-defined type to the input file schema.
+   *
+   * \param [in] name Name of the dictionary
+   * \param [in] description Description of the dictionary
+   *
+   * \return Reference to the created dictionary
+   *****************************************************************************
+   */
+  Table& addStructDictionary(const std::string& name,
```

Comment:
This is just to clarify my understanding. What's the difference between a struct and a dictionary? Can a dictionary have arbitrary members, but a struct only specific ones?

Reply:
This is a good question - I've added a snippet on this to the docs, reproduced here:

Although many of Inlet's input file languages do not distinguish between a "dictionary" type
  and a "record" type, Inlet treats them differently for type safety reasons:

  *Dictionaries* use arbitrary strings or integers for their keys, and their values (entries)
  can only be retreived as a homogenous type.  In other words, dictionaries must map to
  ``std::unordered_map<Key, Value>`` for fixed key and value types.

  *Structs* contain a fixed set of named fields, but these fields can be of any type.
  As the name suggests, these map to ``structs`` in C++.

  In short, if the *key values* vary, use a dictionary.  If the *field types* vary, use a struct.

Reply:
When creating concepts likes this for an interface, even if you document well -- it can be confusing if you attribute new semantics to things people may already have mental models for (Tables, Dictionaries, Structs). 

`if the key values vary, use a dictionary`.  

By vary here you mean the key names aren't validated, where the struct names are? Is that the root of the difference with validating the `names` (or keys) passed when fetching?


An important aspect is if the validation + errors messages for users lead them to understand what is off and make it easy for folks to correct the inputs when things go off the rails, regardless of what constructs you use to do that. 

Hard to understand all that via a PR

Reply:
@cyrush  and i had a conversation about this thread.  We agree there is a confusion between overloaded concepts due to language differences.  We should get these existing PRs in.  Then have a high level discussion to try to minimize the confusion going on.  Then have one PR that addresses our naming conventions.